### PR TITLE
MODINREACH-137 INN-Reach Transactions (All Types): Add get collection endpoint - filter parameters

### PR DESCRIPTION
**Pull request by maryna-steshenko**
## Purpose
Add filter parameters (type, state, central server code, patron agency code, item agency code, patron type, material type) for get all transaction endpoint.
JIRA: [MODINREACH-137](https://issues.folio.org/browse/MODINREACH-137)

## Approach

- InnReachTransaction repository modified.
- InnReachTransaction filter specifications added.
- InnReachTransaction sorting parameters DTO added.
- InnReachTransaction service and controller modified.
- InnReachTransaction controller tests added.
- OpenAPI configuration updated.

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/innreach/specification/InnReachTransactionSpecification.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+      .and(centralCodeIn(parameters.getCentralServerCodes()))
+      .and(patronAgencyIn(parameters.getPatronAgencyCodes()))
+      .and(itemAgencyIn(parameters.getItemAgencyCodes()))
+      .and(patronTypeIn(parameters.getPatronTypes()))
+      .and(centralItemTypeIn(parameters.getCentralItemTypes()));
+  }
+
+  static Specification<InnReachTransaction> isOfType(List<InnReachTransaction.TransactionType> types) {
+    return (transaction, cq, cb) -> {
+      if (Long.class != cq.getResultType()) {
```
Comment (dmtkachenko) [inline src/main/java/org/folio/innreach/specification/InnReachTransactionSpecification.java:34]: Could you please explain why this if statement needed?
My understanding this method should perform comparison of transaction types, like:
`transaction.type IN (<types>)`
and this is done by the following statements

File: src/main/java/org/folio/innreach/specification/InnReachTransactionSpecification.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        var hold = transaction.fetch("hold");
+        hold.fetch("pickupLocation");
+      }
+      if (types == null || types.isEmpty()) {
```
Comment (dmtkachenko) [inline src/main/java/org/folio/innreach/specification/InnReachTransactionSpecification.java]: "if" statement and the following return can be simplified to:
`return CollectionUtils.isEmpty(types) ? cb.conjunction() : transaction.get("type").in(types);`

the same is true for similar cases in the other methods

File: src/main/java/org/folio/innreach/mapper/InnReachTransactionSortingParametersMapper.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+import org.folio.innreach.domain.entity.InnReachTransactionSortingParameters;
+import org.folio.innreach.dto.InnReachTransactionSortingParametersDTO;
+import org.mapstruct.Builder;
+import org.mapstruct.InjectionStrategy;
+import org.mapstruct.Mapper;
+
+@Mapper(componentModel = "spring", injectionStrategy = InjectionStrategy.CONSTRUCTOR, uses = MappingMethods.class, builder = @Builder(disableBuilder = true))
+public interface InnReachTransactionSortingParametersMapper {
+  InnReachTransactionSortingParameters toEntity(InnReachTransactionSortingParametersDTO dto);
```
Comment (dmtkachenko) [inline src/main/java/org/folio/innreach/mapper/InnReachTransactionSortingParametersMapper.java]: do we really need this conversion and `InnReachTransactionSortingParameters` class at all? this is not an entity, right?
looks like the objects are completely identical: have the same fields and field types

File: src/main/resources/swagger.api/schemas/innReachTransactionSortingParametersDTO.json [hunk 4]
```diff
@@ -0,0 +1,80 @@
+{
+  "description": "INN-Reach Transaction sorting parameters",
+  "type": "object",
+  "properties": {
+    "types" : {
+      "description": "List of INN-Reach Transaction types",
+      "type": "array",
+      "items": {
+        "type": "string",
+        "enum" : [
```
Comment (dmtkachenko) [inline src/main/resources/swagger.api/schemas/innReachTransactionSortingParametersDTO.json]: common enum types can be defined in the separate json schema and then reused in many places. in this case the same enum can be reused in innReachTransactionsDTO.json and innReachTransactionSortingParametersDTO.json.

see the following file as an example of enum definition: https://github.com/folio-org/mod-kb-ebsco-java/blob/master/ramls/types/packages/contentTypeEnum.json

File: src/main/resources/swagger.api/schemas/innReachTransactionSortingParametersDTO.json [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+          "LOCAL"
+        ]
+      }
+    },
+    "states" : {
+      "description": "List of INN-Reach Transaction states",
+      "type": "array",
+      "items": {
+        "type": "string",
+        "enum" : [
```
Comment (dmtkachenko) [inline src/main/resources/swagger.api/schemas/innReachTransactionSortingParametersDTO.json]: this enum can be also reused

File: src/main/java/org/folio/innreach/mapper/InnReachTransactionSortingParametersMapper.java (hunk 3 above)
Comment (maryna-steshenko) [reply src/main/java/org/folio/innreach/mapper/InnReachTransactionSortingParametersMapper.java]: This mapper and `InnReachTransactionSortingParameters` entity were added to convert between domain's `TransactionType`/`TransactionState` and DTO's enums.

File: src/main/java/org/folio/innreach/specification/InnReachTransactionSpecification.java (hunk 1 above)
Comment (maryna-steshenko) [reply src/main/java/org/folio/innreach/specification/InnReachTransactionSpecification.java:34]: We need this statement to correctly initialize count query for fetch joining `hold` and `pickupLocation` entities. Wasn't sure, where this check needs to be, so placed it here. Now moved it to a separate specification `fetchHoldAndPickupLocation()`.

**Review by dmtkachenko — APPROVED** [review]

**Review by oliinyko — APPROVED** [review]

**Pull request merged**